docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,490 @@
+      configMap:
+        image: nginx
+        tag: 1.2.0 // updated tag is preserved
+        new-setter: new-setter-value // new setter is pulled
+```
+
+In the above scenario, the value of the setter tag is not updated upstream, so 
+the modified local value will be preserved. But in the following example, both 
+upstream and local values change. So, similar to merging resources, upstream 
+value wins if the same fields in both upstream and local are updated.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-updated.yaml // upstream value changed
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-local.yaml // local value changed
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-updated.yaml // upstream overrides local 
+```
+
+Similarly, the upstream version wins if both upstream and local are updated, else local is preserved.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters.yaml
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.1
+      configPath: setters.yaml
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters-updated.yaml
+```
+
+This might not be what all users expect. But this is the default behavior in case 
+of conflict while merging normal resources as well. In order to provide more visibility to the users, we can add 
+log messages in cases of such conflicts and intimate users about the updated value. 
+In the future, we can add support to a different conflict strategy of `--local-wins`
+as an option to the kpt pkg update command.
+
+#### More examples with expected output
+
+Newly added upstream functions are appended at the end.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+
+Expected output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+    - image: gcr.io/kpt-fn/set-namespace:v0.1
+      configMap:
+        namespace: foo
+    - image: gcr.io/kpt-fn/generate-folders:v0.1
```

Comment:
another expected output could be (to highlight ordering issue):

```
pipeline:
  mutators:
    - image: gcr.io/kpt-fn/apply-setters:v0.1
      configPath: setters.yaml
    - image: gcr.io/kpt-fn/generate-folders:v0.1
    - image: gcr.io/kpt-fn/set-namespace:v0.1
      configMap:
        namespace: foo
```

Reply:
So this ordering is result of `upstream-wins` strategy ? if one select `local-wins`, then will `set-namespace` appear before `generate-folders` ?

Reply:
I'd like to see more about this discussed. It's not clear to me if we always put upstream first or something else?

Reply:
We always put upstream first according to current merge behavior. This is already being followed for merging containers in deployment for example. 

@droot Conflict resolution strategy(local-wins) is something we should discuss in the future and is not in scope for this PR. We are just following the current merge behavior.

Reply:
> @droot Conflict resolution strategy(local-wins) is something we should discuss in the future and is not in scope for this PR. We are just following the current merge behavior.

ok, in that case, we should document the ordering behavior.